### PR TITLE
Miscellaneous fixes for Ocaml warnings.

### PR DESCRIPTION
**Pull request by ejgallego**
Fix some warnings I get. Most of the commits are straighforward, however note the dependency on Ocaml 4.02.3 / Bytes to start fixing some `-safe-string` stuff.

### CODE REVIEW

File: lib/unicode.ml [hunk 1]
```diff
@@ -126,26 +126,29 @@ let utf8_of_unicode n =
...
-      s.[1] <- Char.chr (128 + (n / 64) mod 64);
-      s.[0] <- Char.chr (224 + n / 4096);
-      s
-    end
+    String.init 3 (fun idx ->
+        match idx with
+        | 0 -> Char.chr (224 + n / 4096)
+        | 1 -> Char.chr (128 + (n / 64) mod 64)
+        | 2 -> Char.chr (128 + n mod 64)
+        | _ -> 'x'
```
Comment (maximedenes) [inline lib/unicode.ml]: Shouldn't this be `assert false`? And similarly below?

Comment (silene) [inline lib/unicode.ml]: Actually, if we are willing to use String.init and thus don't care much about performances, we might just as well go the whole way toward higher order. The whole code then becomes:

    if n < 128 then
      String.make 1 (Char.chr n)
    else
      let (m,s) = if n < 2048 then (2,192) else if n < 65536 then (3,224) else (4,240) in
      String.init m (fun i ->
        let j = (n lsr ((m - 1 - i) * 6)) land 63 in
        Char. chr (j + if i = 0 then s else 128))

Comment (ejgallego) [inline lib/unicode.ml]: I am no expert in Unicode and I don't have any preference, (well, I'd rather not have this code in Coq at all but depend on an external library)

I guess that anything that works with immutable strings should be fine for the purposes of this patch.

Comment (herbelin) [inline lib/unicode.ml]: I'm unsure there is a rule allowing to say that depending on an external library is necessarily better. For instance, camomile on which we hesitated to rely a few years ago did not evolve for more than 5 years and their unicode tables are not any more up-to-date.

**Comment by ejgallego** [comment]
Maxime comments addressed. I've also added Guillaume's, feel free to chose any solution.

File: lib/unicode.ml [hunk 2]
```diff
@@ -121,30 +121,44 @@ let classify =
 
 exception End_of_input
 
+(* Guillaume Melquiond proposes this alternative code:
```
Comment (maximedenes) [inline lib/unicode.ml]: After a careful review, this code looks correct indeed. @ejgallego, can you update this PR with Guillaume's code for testing? Thanks!

Comment (ejgallego) [inline lib/unicode.ml]: Ok I will

Comment (ejgallego) [inline lib/unicode.ml]: done

**Pull request merged**